simperby/src/node/state_storage.rs

```diff
@@ -16,13 +12,107 @@ pub enum StateStorageError {
     StorageIntegrityError(String),
 }
 
+type Error = StateStorageError;
+
+impl From<simperby_kv_storage::Error> for Error {
+    fn from(e: simperby_kv_storage::Error) -> Self {
+        Error::StorageError(e)
+    }
+}
+
 impl From<Error> for SimperbyError {
     fn from(e: Error) -> Self {
-        unimplemented!()
+        match e {
+            Error::InvalidOperation(e) => SimperbyError::InvalidOperation(e),
+            Error::StorageError(e) => SimperbyError::StorageError(e),
+            Error::StorageIntegrityError(e) => SimperbyError::StorageIntegrityError(e),
+        }
     }
 }
 
-type Error = StateStorageError;
+/// Reads an essential (existence-guaranteed) item from the storage.
+async fn get_json_from_storage<T>(storage: &dyn KVStorage, key: Hash256) -> Result<T, Error>
+where
+    T: serde::de::DeserializeOwned,
+{
+    let value = match storage.get(key.clone()).await {
+        Ok(x) => x,
+        Err(simperby_kv_storage::Error::NotFound) => {
+            return Err(Error::StorageIntegrityError(format!(
+                "missing essential item in the state stroage: {}",
+                key
+            )))
+        }
+        Err(e) => return Err(e.into()),
+    };
+    serde_json::from_slice(&value).map_err(|e| Error::StorageIntegrityError(e.to_string()))
+}
+
+async fn get_hash_from_storage(storage: &dyn KVStorage, key: Hash256) -> Result<Hash256, Error> {
+    let value = match storage.get(key.clone()).await {
+        Ok(x) => x,
+        Err(simperby_kv_storage::Error::NotFound) => {
+            return Err(Error::StorageIntegrityError(format!(
+                "missing essential item in the state stroage: {}",
+                key
+            )))
+        }
+        Err(e) => return Err(e.into()),
+    };
+    let msg = if value.len() > 1024 {
+        format!(
+            "item hash is corrupted: (too long; emited; {})",
+            value.len()
+        )
+    } else {
+        format!("item hash is corrupted: {:?}", value)
+    };
+    let data: [u8; 32] = TryInto::try_into(value).map_err(|_| Error::StorageIntegrityError(msg))?;
+    Ok(Hash256::from_array(data))
+}
+
+async fn insert_or_update_json_to_storage<T>(
+    storage: &mut dyn KVStorage,
+    key: Hash256,
+    value: &T,
+) -> Result<(), Error>
+where
+    T: serde::Serialize,
+{
+    let value = serde_json::to_vec(value).unwrap();
+    let hash = Hash256::hash(&value);
+    storage.insert_or_update(key.clone(), &value).await?;
+    storage
+        .insert_or_update(create_hash_key(key), hash.as_ref())
+        .await?;
```

Comment:
Why do we need to store key's hash value in state storage?

Reply:
Because it will be frequently queried. Later this will be converted into a Merkle node. (i.e., this is an inevitable overhead)